### PR TITLE
Wire actor

### PR DESCRIPTION
**Pull request by pawelpanasewicz**
#108 

PR introduces 3 macros available after importing `com.softwaremill.macwire.akkasupport._`. The macros are:

```scala
wireProps[SomeActor] //returns Props
wireAnonymousActor[SomeActor] //returns ActorRef
wireActor[SomeActor]("actorName") //returns ActorRef
```
All dependencies of prmiary public constructor of `SomeActor` must be in scope. As well `ActorRefFactory` (for example `ActorSystem` or `ActorContext`) must be in scope.

Please see the `Demo` object (last commit) to familiarize with usage.

### CODE REVIEW

**Comment by pawelpanasewicz** [comment]
The build has failed on Travis. 

Travis uses java**7**, however if you have java**8** on path build works. It is strange for me

```» ./sbt -2.11 -v clean test                                                                                                                /cygdrive/d/devstation-workspace/projects/macwire
[residual] arg = 'clean'
[residual] arg = 'test'
No extra sbt options have been defined
Detected sbt version 0.13.12
Using default jvm options
Detected Java version: 1.7.0_79
# Executing command line:
java
-XX:MaxPermSize=384m
-Xms512m
-Xmx1536m
-Xss2m
-jar
D:\devstation-workspace\bin\cygwin\home\W541\.sbt\launchers\0.13.12\sbt-launch.jar
-2.11
clean
test
...
[info] CompileTests:
[info] wireActor-1-simple.success
[info] - should compile & run
[info] anonFuncArgsWiredOk.success
[info] - should compile & run
...
[info] simpleValsMissingValue.failure
[info] - should cause a compile error
[info] Run completed in 1 minute, 35 seconds.
[info] Total number of tests run: 72
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 72, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 117 s, completed 31-Dec-2016 16:11:33
```

File: macrosAkkaTests/src/test/resources/test-cases-TODO/wireActor-8-dependencyByName.success [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+val a = new A {}
+
+val system = ActorSystem("wireActor-8-dependencyByName.success")
+
+val someActor = wireActor[SomeActor]("bob")
+
+try {
+  someActor ! "Hey someActor"
+} finally system.terminate()
```
Comment (ktoso) [inline macrosAkkaTests/src/test/resources/test-cases-TODO/wireActor-8-dependencyByName.success]: why the try finally?

**Comment by ktoso** [comment]
The move `  .../src/test/scala/javax/inject/Inject.scala → .../src/main/scala/javax/inject/Inject.scala` seems incorrect IMO. Why would you provide the javax inject annotation? It should come as dependency if people need it.

File: macrosAkkaTests/src/test/resources/test-cases-TODO/wireActor-8-dependencyByName.success (hunk 1 above)
Comment (pawelpanasewicz) [inline macrosAkkaTests/src/test/resources/test-cases-TODO/wireActor-8-dependencyByName.success]: Indeed, there is no good reason for it. I'll remove it.

**Comment by pawelpanasewicz** [comment]
@ktoso , In terms of: 
>    .../src/main/scala/javax/inject/Inject.scala

This file is part of `test-util` project which is base for all tests projects. It's not published but used only during tests. I've moved it from `tests\src\test\scala\javax\inject\` into `test-util\src\main\javax\inject`. In this PR I've just followed the conventions and existing state of code base and focused on `wireActor` stuff. 

However, I agree that this should come as dependency. I am going to address this issue and several other concerns in my next PRs :)

File: macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.softwaremill.macwire
+package macwireakka
```
Comment (adamw) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala]: Maybe we could simplify the package name to just `com.softwaremill.macwire.akka`? Not sure if there's any sense in repeating the macwire part

File: macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  import c.universe._
+  lazy val targetType: Type = implicitly[c.WeakTypeTag[T]].tpe
+  lazy val dependencyResolver = new DependencyResolver[c.type](c, log)
+
+  lazy val targetConstructor: Symbol = log.withBlock(s"Looking for targetConstructor for $targetType"){
+    val publicConstructors = targetType.members
+      .filter(m => m.isMethod && m.asMethod.isConstructor && m.isPublic)
+      .filterNot(isPhantomConstructor)
+    log.withBlock(s"There are ${publicConstructors.size} eligible constructors" ) { publicConstructors.foreach(c => log(showConstructor(c))) }
+    val isInjectAnnotation = (a: Annotation) => a.toString == "javax.inject.Inject"
```
Comment (adamw) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala]: The code here seems to have quite a lot in common with the one in `MacwireMacros`. Maybe it would make sense to extract some common parts?

File: macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    *
+    *  //res1: Iterable[String] = List(X.<init>, A.$init$)
+    *  }}}
+    *
+    *  The {{{A.$init$}}} is the phantom constructor and we don't want it.
+    *
+    *  In other words, if we don't filter such constructor using this function
+    *  'wireActor-12-noPublicConstructor.failure' will compile and throw exception during runtime but we want to fail it during compilation time.
+    */
+  private def isPhantomConstructor(constructor: Symbol): Boolean = constructor.asMethod.fullName.endsWith("$init$")
```
Comment (adamw) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala]: Also, the "normal" `wire` can probably benefit from phantom constructor filtering as well?

File: macrosAkkaTests/src/test/resources/test-cases/wireActor-12-noPublicConstructor.failure [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import akka.actor.{Actor, ActorSystem, Props}
+import com.softwaremill.macwire.macwireakka._
+
+/**
+  * There are to many constructors annotated with @Inject
```
Comment (adamw) [inline macrosAkkaTests/src/test/resources/test-cases/wireActor-12-noPublicConstructor.failure]: wrong comment?

File: macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala (hunk 2 above)
Comment (pawelpanasewicz) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala]: I've intentionally named it differently than 'akka' in order to bypass the problem when you need to uglyfy _akka_ imports with `_root_` prefix when you import it after importing `com.softwaremill.macwire.akka`.

Consider such code:

```scala
package whatever

import com.softwaremill.macwire._
import com.softwaremill.macwire.akka._ 

import akka.actor.ActorSystem

object Demo extends App {

  val system = ActorSystem("Demo")
//...
}

//Error:(8, 13) object actor is not a member of package com.softwaremill.macwire.akka
//import akka.actor.ActorSystem

//Error:(10, 16) not found: value ActorSystem
//  val system = ActorSystem("Demo")

```

This will not compile because compiler will look for `ActorSystem` in macwire. For begginers it can be confusing. Especially when users rely on IDE support and for example try to fix imports from Intellij which will not resolve it properly.

Of course we can simply fix it by adding `import akka.actor.ActorSystem` in front _macwire_ or `import _root_.akka.actor.ActorSystem` somewhere after _macwire_.

I was considering as well placing everything in `com.softwaremill.macwireakka` but in such scenario private stuff in `com.softwaremill.macwire` would not be accessible in this project.

What do you think? Shall we rename it?

File: macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala (hunk 3 above)
Comment (pawelpanasewicz) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala]: Sure! This will require modularize and refactor a lill bit of the 'macros' project. I can do it :)

File: macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala (hunk 4 above)
Comment (pawelpanasewicz) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala]: Sure! I can do it :)

File: macrosAkkaTests/src/test/resources/test-cases/wireActor-12-noPublicConstructor.failure (hunk 5 above)
Comment (pawelpanasewicz) [inline macrosAkkaTests/src/test/resources/test-cases/wireActor-12-noPublicConstructor.failure]: Ooops, Let me fix it :)

**Comment by pawelpanasewicz** [comment]
> And there's a lot of tests :) 

There was no other option ;) Thanks to that few bugs were found in development phase and not in production. Moreover we will benefit from them every time when we change the code.

> One thing that would be great to have is also a new section in the README. 

I was thinking of _README update_ as seperate PR. It can be merged just after release, so it will help keeping docs up to date.

> And sorry for the late review :)

That's fine :)

File: macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala (hunk 3 above)
Comment (adamw) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/internals/Crimper.scala]: thanks :)

File: macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala (hunk 2 above)
Comment (adamw) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala]: Ah, didn't think about that, good catch :)

I don't have any good ideas about that really ... maybe `akkaActors`? But then it's the double-a ;)

Comment (pawelpanasewicz) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala]: How about `akkaSupport`? Hmm, still double kk and pp ;)

Comment (adamw) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala]: ah, better! :) just name it `akkasupport` to avoid problems with FSs & case-sensititity ;)

Comment (pawelpanasewicz) [inline macrosAkka/src/main/scala/com/softwaremill/macwire/macwireakka/MacwireAkkaMacros.scala]: sure :)

**Comment by pawelpanasewicz** [comment]
In terms of code for 'wireActor' - this is it :) Please read and review. I terms of couple lines of docs - hey will come in separate PR.

File: macrosAkkaTests/src/test/scala/com/softwaremill/macwire/akkasupport/AfterAllTerminate.scala [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package com.softwaremill.macwire.akkasupport
+
+import akka.actor.ActorSystem
+import akka.dispatch.MonitorableThreadFactory
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+object AfterAllTerminate {
```
Comment (adamw) [inline macrosAkkaTests/src/test/scala/com/softwaremill/macwire/akkasupport/AfterAllTerminate.scala]: maybe add some explanation why this is needed?

File: macros/src/main/scala/com/softwaremill/macwire/internals/CompanionCrimper.scala [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    as.foreach(x => log.withBlock(s"There are ${x.size} apply methods:" ) { x.foreach(c => log(showApply(c))) })
+    as
+  }
+
+  lazy val apply: Option[Symbol] = applies.flatMap( _ match {
+    case applyMethod :: Nil => Some(applyMethod)
+    case _ => None
+  })
+
+  lazy val applySelect: Option[Select] = apply.map(a => Select(Ident(targetType.typeSymbol.companion), a))
```
Comment (adamw) [inline macros/src/main/scala/com/softwaremill/macwire/internals/CompanionCrimper.scala]: I don't quite understand the idea with all these `lazy val`s :) Why are they lazy and public? They make the code quite hard to follow ... there's a lot of jumping around. Maybe this could be simplified, and when possible made private?

File: macrosAkkaTests/src/test/scala/com/softwaremill/macwire/akkasupport/AfterAllTerminate.scala (hunk 6 above)
Comment (pawelpanasewicz) [inline macrosAkkaTests/src/test/scala/com/softwaremill/macwire/akkasupport/AfterAllTerminate.scala]: mkay

File: macros/src/main/scala/com/softwaremill/macwire/internals/CompanionCrimper.scala (hunk 7 above)
Comment (pawelpanasewicz) [inline macros/src/main/scala/com/softwaremill/macwire/internals/CompanionCrimper.scala]: Consider this usage of `lazy val`s:

```scala
object MacwireMacros { ...
  def wire_impl[T: c.WeakTypeTag](c: blackbox.Context): c.Expr[T] = {
    import c.universe._

    val constructorCrimper = new ConstructorCrimper[c.type, T](c, log)
    val companionCrimper = new CompanionCrimper[c.type, T](c, log)

    lazy val targetType = companionCrimper.targetType.toString
    lazy val whatWasWrong: String = {
      if(constructorCrimper.constructor.isEmpty && companionCrimper.companionType.isEmpty)
        s"Cannot find a public constructor nor a companion object for [$targetType]"
      else if(companionCrimper.applies.isDefined && companionCrimper.applies.get.isEmpty)
        s"Companion object for [$targetType] has no apply methods constructing target type."
      else if(companionCrimper.applies.isDefined && companionCrimper.applies.get.size > 1)
        s"No public primary constructor found for $targetType and multiple matching apply methods in its companion object were found."
      else s"Target type not supported for wiring: $targetType. Please file a bug report with your use-case."
    }

    val code: Tree = (constructorCrimper.constructorTree orElse companionCrimper.applyTree) getOrElse
      c.abort(c.enclosingPosition, whatWasWrong)
    log(s"Generated code: ${showCode(code)}, ${showRaw(code)}")
    c.Expr(code)
  }
```

`companionCrimper.applyTree` is a `lazy val`. That means that if `constructorCrimper.constructorTree` is `Some` it won't be computed - faster compilation time.

You may ask then why the `constructorCrimper.constructorTree` is lazy? It seems that it is computed always. Well, not exactly. In `Macros.wire_impl` - yes it is. But `MacwireAkkaMacros` uses (indirectly) `ConstructorCrimper` as well but not this particular lazy val. If we made `constructorCrimper.constructorTree` eager `val` it would be computed even if we don't used it.

Ok. So do we really need to have all `val`s in this [Crimper](https://www.google.pl/search?q=crimper&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiTrt_kke3RAhXIbZoKHQjcAx4Q_AUICCgB&biw=1920&bih=1070)-things to be lazy? Actually no, we don't. However it's easier to make them all lazy than manually track what is needed and when. Can we have some of them private and some not? Yes - but sometimes it's easier to type `companionCrimper. [ctrl+space]` and with IDE support see what's inside and pick up what is needed. 

Btw. Lazy vals aren't publicly exposed. They are visible only in macwire` package (see `private[macwire] class ConstructorCrimper ...`).

Does it make any sense now?

Comment (adamw) [inline macros/src/main/scala/com/softwaremill/macwire/internals/CompanionCrimper.scala]: ok, let's see how it works in practice :)

**Comment by adamw** [comment]
Released 2.3.0 - thanks a lot for your work!
Could you take a look at https://github.com/adamw/macwire#akka-integration if I got everything fine?

**Comment by adamw** [comment]
@pawelpanasewicz ^ :) (forgot the mention)

**Comment by pawelpanasewicz** [comment]
Hey @adamw! Cool, we have released :) I can't wait to start using this in my projects ;)
As promised, there will be one more PR with docs. Yes, I'll take a look at your chapter.
